File: datalib/src/main/java/nl/acidcats/tumblrlikes/core/usecases/photos/UpdatePhotoPropertyUseCaseImpl.java
package nl.acidcats.tumblrlikes.core.usecases.photos;

import javax.inject.Inject;

import nl.acidcats.tumblrlikes.core.models.Photo;
import nl.acidcats.tumblrlikes.core.repositories.PhotoDataRepository;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created on 03/10/2018.
 */
public class UpdatePhotoPropertyUseCaseImpl implements UpdatePhotoPropertyUseCase {

    private PhotoDataRepository _photoDataRepository;

    @Inject
    public UpdatePhotoPropertyUseCaseImpl(PhotoDataRepository photoDataRepository) {
        _photoDataRepository = photoDataRepository;
    }

    @Override
    public Observable<Photo> updateLike(long id, boolean isLiked) {
        return Observable
                .fromCallable(() -> {
                    _photoDataRepository.setPhotoLiked(id, isLiked);

                    return _photoDataRepository.getPhotoById(id);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public Observable<Photo> updateFavorite(long id, boolean isFavorite) {
        return Observable
                .fromCallable(() -> {
                    _photoDataRepository.setPhotoFavorite(id, isFavorite);

                    if (isFavorite) {
                        _photoDataRepository.setPhotoLiked(id, true);
                    }

                    return _photoDataRepository.getPhotoById(id);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public Observable<Photo> setHidden(long id) {
        return Observable
                .fromCallable(() -> {
                    _photoDataRepository.setPhotoFavorite(id, false);
                    _photoDataRepository.setPhotoLiked(id, false);

                    _photoDataRepository.hidePhoto(id);

                    return _photoDataRepository.getPhotoById(id);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: datalib/src/main/kotlin/nl/acidcats/tumblrlikes/core/usecases/photos/UpdatePhotoPropertyUseCaseImpl.kt
package nl.acidcats.tumblrlikes.core.usecases.photos

import nl.acidcats.tumblrlikes.core.models.Photo
import nl.acidcats.tumblrlikes.core.repositories.PhotoDataRepository
import nl.acidcats.tumblrlikes.core.viewmodels.PhotoViewModel
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created on 24/10/2018.
 */
class UpdatePhotoPropertyUseCaseImpl @Inject constructor(private val photoDataRepository: PhotoDataRepository) : UpdatePhotoPropertyUseCase {
    override fun updateLike(id: Long, isLiked: Boolean): Observable<PhotoViewModel> {
        return applyUpdate(id) {
            photoDataRepository.setPhotoLiked(id, isLiked)

            if (!isLiked) {
                photoDataRepository.setPhotoFavorite(id, false)
            }
        }
    }

    override fun updateFavorite(id: Long, isFavorite: Boolean): Observable<PhotoViewModel> {
        return applyUpdate(id) {
            photoDataRepository.setPhotoFavorite(id, isFavorite)

            if (isFavorite) {
                photoDataRepository.setPhotoLiked(id, true)
            }
        }
    }

    override fun setHidden(id: Long): Observable<PhotoViewModel> {
        return applyUpdate(id) {
            photoDataRepository.setPhotoLiked(id, false)
            photoDataRepository.setPhotoFavorite(id, false)

            photoDataRepository.hidePhoto(id)
        }
    }

    fun applyUpdate(id: Long, updateProperty: () -> Unit): Observable<PhotoViewModel> {
        return Observable.fromCallable { updateProperty() }
                .map { photoDataRepository.getPhotoById(id) }
                .map { PhotoViewModel.from(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }
}